### PR TITLE
Feature/timeline

### PR DESCRIPTION
**Pull request by abdulhakim2902**

### CODE REVIEW

**Comment by rei-gun** [comment]
pretty good commit messages. just remember to use past tense e.g. `removed` instead of `remove`

File: src/services/experience.service.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+export class ExperienceService {
+  constructor (
+    @repository(SavedExperienceRepository)
+    protected savedExperienceRepository: SavedExperienceRepository,
+    @repository(ExperienceRepository)
+    protected experienceRepository: ExperienceRepository,
+  ) {}
+
+  async getSelectedExperience(userId: string): Promise<Experience | null> {
```
Comment (rei-gun) [inline src/services/experience.service.ts:13]:
```suggestion
  async getExperience(userId: string): Promise<Experience | null> {
```
bisa di shorten nama function nya

File: src/services/experience.service.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    const selectedUserSavedExperience = await this.savedExperienceRepository.findOne({
```
Comment (rei-gun) [inline src/services/experience.service.ts:14]:
```suggestion
    const experience = await this.savedExperienceRepository.findOne({
```
bisa di shorten

**Review by rei-gun — COMMENTED** [review]
every time you want to use `found` in a variable name, ask yourself if it's really necessary. because you usually do a check `if (!foundExperience) {}` and it's already readable that the item was not found if it goes into this statement

File: src/services/filter.service.ts [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    @service(TagService)
+    protected tagService: TagService,
+    @service(FriendService)
+    protected friendService: FriendService
+  ) {}
+
+  async filterByExperience(userId: string): Promise<Where | null> {
+    const foundSelectedExperience = await this.experienceService.getSelectedExperience(userId);
+
+    if (!foundSelectedExperience) return null;
```
Comment (rei-gun) [inline src/services/filter.service.ts]: ask yourself if the `found` is necessary in the variable names here

**Review by rei-gun — APPROVED** [review]

File: src/controllers/transaction.controller.ts [hunk 4]
```diff
@@ -65,27 +57,22 @@ export class TransactionController {
     const {from, to, value, cryptocurrencyId: cryptoId, postId} = transaction;
```
Comment (irmannmal) [inline src/controllers/transaction.controller.ts:57]: for consistency change `cryptocurrencyId: cryptoId` to `cryptoCurrencyId`

File: src/controllers/transaction.controller.ts [hunk 5]
```diff
@@ -65,27 +57,22 @@ export class TransactionController {
...
 
     // Validate if crypto already exist
-    await this.cryptocurrencyRepository.findById(
-      transaction.cryptocurrencyId.toUpperCase(),
-    );
+    await this.cryptocurrencyRepository.findById(transaction.cryptocurrencyId.toUpperCase());
```
Comment (irmannmal) [inline src/controllers/transaction.controller.ts:60]: use camelCase
from `cryptocurrencyRepository` to `cryptoCurrencyRepository`

File: src/controllers/transaction.controller.ts [hunk 6]
```diff
@@ -65,27 +57,22 @@ export class TransactionController {
...
-      await this.transactionService.isTotalTipInPersonUpdated(
-        transaction.to,
-        postId,
-        cryptoId,
-        value,
-      );
+    const isPeopleTipUpdated = await this.transactionService.isTotalTipInPersonUpdated(
+      transaction.to,
+      postId,
+      cryptoId,
```
Comment (irmannmal) [inline src/controllers/transaction.controller.ts:70]: should be `cryptoCurrencyId`

File: src/controllers/transaction.controller.ts [hunk 7]
```diff
@@ -65,27 +57,22 @@ export class TransactionController {
...
+      value,
+    );
 
     // Reward to FROM for doing transactions
-    this.cryptocurrencyService.sendMyriadReward(
-      transaction.from,
-    ) as Promise<void>;
+    this.cryptocurrencyService.sendMyriadReward(transaction.from) as Promise<void>;
```
Comment (irmannmal) [inline src/controllers/transaction.controller.ts:75]: use camel case
from `cryptocurrencyService` to `cryptoCurrencyService`

File: src/enums/timelinefilter-type.enum.ts [hunk 8]
```diff
@@ -17,3 +17,10 @@ export enum DefaultInq {
   TAGS = 'cryptocurrency,blockchain,technology',
   PEOPLE = 'elonmusk,gavofyork,W3F_Bill,CryptoChief,BillGates,vitalikbuterineth',
 }
+
+export enum TimelineType {
+  EXPERIENCE = 'experience',
+  TRENDING = 'trending',
```
Comment (irmannmal) [inline src/enums/timelinefilter-type.enum.ts:23]: reorder enum, `ALL` should on the top

File: src/initDatabase.ts [hunk 9]
```diff
@@ -140,8 +135,7 @@ export class InitDatabase extends BootMixin(
         } else {
           const oneDay: number = 60 * 60 * 24 * 1000;
           const isOneDay: boolean =
-            new Date().getTime() - new Date(foundTag.updatedAt).getTime() >
-            oneDay;
+            new Date().getTime() - new Date(foundTag.updatedAt).getTime() > oneDay;
```
Comment (irmannmal) [inline src/initDatabase.ts:138]: make this function as `DateUtils`

File: src/initDatabase.ts [hunk 10]
```diff
@@ -156,43 +150,23 @@ export class InitDatabase extends BootMixin(
...
     const userRepository = await this.getRepository(UserRepository);
-    const savedExperienceRepository = await this.getRepository(
-      SavedExperienceRepository,
-    );
+    const savedExperienceRepository = await this.getRepository(SavedExperienceRepository);
     const experienceRepository = await this.getRepository(ExperienceRepository);
-    const userCredRepository = await this.getRepository(
-      UserCredentialRepository,
-    );
+    const userCredRepository = await this.getRepository(UserCredentialRepository);
```
Comment (irmannmal) [inline src/initDatabase.ts:157]: change this to `userCredentialRepository`

File: src/keys.ts [hunk 11]
```diff
@@ -33,13 +31,11 @@ export namespace RefreshTokenConstants {
   /**
    * The default secret used when generating refresh token.
    */
-  export const REFRESH_SECRET_VALUE =
-    process.env.REFRESH_TOKEN_SECRET_KEY ?? 'r3fr35htok3n';
+  export const REFRESH_SECRET_VALUE = process.env.REFRESH_TOKEN_SECRET_KEY ?? 'r3fr35htok3n';
```
Comment (irmannmal) [inline src/keys.ts:34]: for audit compliance, default secret should be `'secret'`

File: src/models/people.model.ts [hunk 12]
```diff
@@ -9,7 +9,7 @@ import {PersonTip} from './person-tip.model';
     mongodb: {
       collection: 'people',
     },
-    hiddenProperties: ['totalTips'],
+    // TODO: removed unused line
```
Comment (irmannmal) [inline src/models/people.model.ts:12]: remove this line, handled by git history

File: src/models/people.model.ts [hunk 13]
```diff
@@ -53,11 +53,7 @@ export class People extends Entity {
   })
   profileImageURL: string;
 
-  @property({
-    type: 'boolean',
-    default: false,
-  })
-  hide?: boolean;
+  // TODO: removed unused field (hide property)
```
Comment (irmannmal) [inline src/models/people.model.ts:56]: remove this line, handled by git history

File: src/models/tag.model.ts [hunk 14]
```diff
@@ -1,19 +1,12 @@
...
       collection: 'tags',
     },
-    scope: {
-      order: [
-        `${OrderFieldType.COUNT} ${OrderType.DESC}`,
-        `${OrderFieldType.UPDATEDAT} ${OrderType.DESC}`,
-      ],
-      limit: 10,
-    },
+    // TODO: Remove default filter
```
Comment (irmannmal) [inline src/models/tag.model.ts:9]: remove this line, handled by git history

File: src/models/tag.model.ts [hunk 15]
```diff
@@ -29,11 +22,7 @@ export class Tag extends Entity {
   })
   id: string;
 
-  @property({
-    type: 'boolean',
-    default: false,
-  })
-  hide: boolean;
+  // TODO: Remove unused field (hide property)
```
Comment (irmannmal) [inline src/models/tag.model.ts:25]: remove this line, handled by git history

**Review by irmannmal — CHANGES_REQUESTED** [review]

File: src/controllers/transaction.controller.ts (hunk 5 above)
Comment (abdulhakim2902) [reply src/controllers/transaction.controller.ts:60]: cryptocurrency bukannya 1 kata ya, bukan 2 kata yang terpisah

File: src/enums/timelinefilter-type.enum.ts (hunk 8 above)
Comment (irmannmal) [reply src/enums/timelinefilter-type.enum.ts:23]: bor @abdulhakim2902 please solve this

File: src/helpers/date-utils.ts [hunk 16]
```diff
@@ -0,0 +1,20 @@
+export class DateUtils {
+  second: number;
```
Comment (irmannmal) [inline src/helpers/date-utils.ts:2]: this should be final variable, do not put on constructor,
and for consistency, should be like this

this.second = 1000;
this.minute = 60 * second;
this.hour = 60 * minute;
this.day = 24 * hour;
this.year = 365 * day;

**Review by irmannmal — CHANGES_REQUESTED** [review]

File: src/helpers/date-utils.ts (hunk 16 above)
Comment (abdulhakim2902) [reply src/helpers/date-utils.ts:2]: Sip, done ya..

File: src/services/tag.service.ts [hunk 17]
```diff
@@ -34,16 +37,26 @@ export class TagService {
           updatedAt: new Date().toString(),
         }) as Promise<Tag>;
       } else {
-        const oneDay: number = 60 * 60 * 24 * 1000;
-        const isOneDay: boolean =
-          new Date().getTime() - new Date(foundTag.updatedAt).getTime() >
-          oneDay;
+        const oneDay = day;
+        const isToday = today(foundTag.updatedAt) > oneDay;
```
Comment (irmannmal) [inline src/services/tag.service.ts:41]: move this function to DateUtils,
this line should be like this

`const isToday = today(foundTag.updatedAt);`

File: src/helpers/date-utils.ts [hunk 18]
```diff
@@ -0,0 +1,20 @@
...
+  constructor() {
+    // In miliseconds
+    this.second = 1000;
+    this.minute = 60 * this.second;
+    this.hour = 60 * this.minute;
+    this.day = 24 * this.hour;
+    this.year = 365 * this.day;
+  }
+
+  today(date: string): number {
```
Comment (irmannmal) [inline src/helpers/date-utils.ts:17]: isToday(date: string): bool {
 return (new Date().getTime() - new Date(date).getTime()) > this.day;
}

**Review by irmannmal — CHANGES_REQUESTED** [review]

**Review by irmannmal — APPROVED** [review]

**Pull request merged**